File: make-mcp/src/scripts/import-modules.ts
#!/usr/bin/env node
/**
 * Import Modules Script
 * 
 * Reads modules from data/modules-input.json and imports them into the database.
 * 
 * Usage:
 *   npm run import-modules
 *   
 * Or with custom file:
 *   node dist/scripts/import-modules.js path/to/your-modules.json
 */

import fs from 'fs';
import path from 'path';
import { ModuleRepository } from '../database/module-repository';

interface ModuleParameter {
    name: string;
    type: string;
    required: boolean;
    description?: string;
    options?: string[];
    default_value?: any;
}

interface ModuleInput {
    id: string;
    name: string;
    app: string;
    type: 'trigger' | 'action' | 'search' | 'router' | 'aggregator' | 'transformer' | 'iterator' | 'repeater';
    description: string;
    category: string;
    parameters: ModuleParameter[];
    icon_url?: string;
    documentation_url?: string;
    is_premium?: boolean;
}

interface ModulesInput {
    modules: ModuleInput[];
}

class ModuleImporter {
    private repository: ModuleRepository;
    private stats = {
        total: 0,
        imported: 0,
        updated: 0,
        failed: 0,
        errors: [] as string[]
    };

    constructor() {
        this.repository = new ModuleRepository();
    }

    /**
     * Import modules from JSON file
     */
    async importFromFile(filePath: string): Promise<void> {
        console.log(`\n🔄 Starting module import from: ${filePath}\n`);

        // Read and parse file
        let data: ModulesInput;
        try {
            const fileContent = fs.readFileSync(filePath, 'utf-8');
            data = JSON.parse(fileContent);
        } catch (error) {
            console.error('❌ Failed to read or parse file:', error);
            throw error;
        }

        if (!data.modules || !Array.isArray(data.modules)) {
            throw new Error('Invalid format: expected { modules: [...] }');
        }

        this.stats.total = data.modules.length;
        console.log(`📊 Found ${this.stats.total} modules to import\n`);

        // Import each module
        for (let i = 0; i < data.modules.length; i++) {
            const module = data.modules[i];
            console.log(`[${i + 1}/${this.stats.total}] Processing: ${module.name} (${module.id})`);

            try {
                await this.importModule(module);
                this.stats.imported++;
                console.log(`  ✅ Imported successfully\n`);
            } catch (error) {
                this.stats.failed++;
                const errorMsg = `Failed to import ${module.id}: ${error}`;
                this.stats.errors.push(errorMsg);
                console.error(`  ❌ ${errorMsg}\n`);
            }
        }

        // Print summary
        this.printSummary();
    }

    /**
     * Import a single module
     */
    private async importModule(moduleInput: ModuleInput): Promise<void> {
        // Validate required fields
        this.validateModule(moduleInput);

        // Check if module already exists
        const existingModule = this.repository.getModuleByName(moduleInput.id);
        const isUpdate = !!existingModule;

        // Insert or update module
        const moduleId = this.repository.upsertModule({
            module_name: moduleInput.id,
            module_type: moduleInput.type,
            app_name: moduleInput.app,
            app_slug: this.slugify(moduleInput.app),
            description: moduleInput.description,
            documentation: this.generateDocumentation(moduleInput),
            parameters: this.convertParameters(moduleInput.parameters),
            examples: this.generateExamples(moduleInput),
            category: moduleInput.category,
            popularity_score: this.calculatePopularityScore(moduleInput),
            is_premium: moduleInput.is_premium || false
        });

        if (isUpdate) {
            this.stats.updated++;
            console.log(`  ℹ️  Updated existing module`);
        }

        console.log(`  📝 Imported ${moduleInput.parameters.length} parameters`);
    }

    /**
     * Validate module input
     */
    private validateModule(module: ModuleInput): void {
        if (!module.id) throw new Error('Module ID is required');
        if (!module.name) throw new Error('Module name is required');
        if (!module.app) throw new Error('Module app is required');
        if (!module.type) throw new Error('Module type is required');
        if (!['trigger', 'action', 'search', 'router', 'aggregator', 'transformer', 'iterator', 'repeater'].includes(module.type)) {
            throw new Error(`Invalid module type: ${module.type}`);
        }
        if (!module.parameters) {
            module.parameters = [];
        }
        if (!Array.isArray(module.parameters)) {
            throw new Error('Parameters must be an array');
        }
    }

    /**
     * Convert parameters to repository format
     */
    private convertParameters(params: ModuleParameter[]): any[] {
        return params.map(param => ({
            name: param.name,
            type: param.type,
            required: param.required,
            description: param.description || '',
            default_value: param.default_value,
            options: param.options
        }));
    }

    /**
     * Generate documentation
     */
    private generateDocumentation(module: ModuleInput): string {
        const paramDocs = module.parameters
            .map(p => {
                const required = p.required ? '(required)' : '(optional)';
                const options = p.options ? ` Options: ${p.options.join(', ')}` : '';
                return `- **${p.name}** ${required}: ${p.description || 'No description'}${options}`;
            })
            .join('\n');

        return `
# ${module.name}

${module.description}

## App
${module.app}

## Type
${module.type}

## Parameters
${paramDocs || 'No parameters'}

## Category
${module.category}
        `.trim();
    }

    /**
     * Generate example configurations
     */
    private generateExamples(module: ModuleInput): any[] {
        const exampleParams: any = {};
        
        module.parameters.forEach(param => {
            if (param.required) {
                exampleParams[param.name] = param.default_value || this.getExampleValue(param);
            }
        });

        return [{
            description: `Example: ${module.name}`,
            parameters: exampleParams
        }];
    }

    /**
     * Get example value for parameter type
     */
    private getExampleValue(param: ModuleParameter): any {
        if (param.options && param.options.length > 0) {
            return param.options[0];
        }
        
        switch (param.type) {
            case 'text':
                return `example_${param.name}`;
            case 'number':
                return 1;
            case 'boolean':
                return true;
            case 'array':
                return [];
            case 'object':
                return {};
            case 'select':
                return param.options?.[0] || 'option1';
            default:
                return `example_${param.name}`;
        }
    }

    /**
     * Calculate popularity score based on heuristics
     */
    private calculatePopularityScore(module: ModuleInput): number {
        let score = 50; // Base score

        // Core modules get higher scores
        if (['HTTP', 'Webhooks', 'Tools'].includes(module.app)) {
            score += 30;
        }

        // Popular apps get bonus
        const popularApps = ['Google Sheets', 'Gmail', 'Slack', 'Airtable', 'Notion'];
        if (popularApps.includes(module.app)) {
            score += 20;
        }

        // Triggers are often more important than actions
        if (module.type === 'trigger') {
            score += 10;
        }

        return Math.min(100, score);
    }

    /**
     * Convert string to slug
     */
    private slugify(text: string): string {
        return text
            .toLowerCase()
            .replace(/[^\w\s-]/g, '')
            .replace(/\s+/g, '-')
            .replace(/-+/g, '-')
            .trim();
    }

    /**
     * Print import summary
     */
    private printSummary(): void {
        console.log('\n' + '='.repeat(60));
        console.log('📊 IMPORT SUMMARY');
        console.log('='.repeat(60));
        console.log(`Total modules:     ${this.stats.total}`);
        console.log(`✅ Imported:       ${this.stats.imported}`);
        console.log(`🔄 Updated:        ${this.stats.updated}`);
        console.log(`❌ Failed:         ${this.stats.failed}`);
        console.log('='.repeat(60));

        if (this.stats.errors.length > 0) {
            console.log('\n❌ ERRORS:');
            this.stats.errors.forEach(error => console.log(`  - ${error}`));
        }

        console.log('\n✨ Import complete!\n');

        // Print database statistics
        const dbStats = this.repository.getStats();
        console.log('📈 DATABASE STATISTICS:');
        console.log(`  Total modules:   ${dbStats.total_modules}`);
        console.log(`  Total apps:      ${dbStats.total_apps}`);
        console.log(`  Total params:    ${dbStats.total_parameters}`);
        console.log(`  Categories:      ${dbStats.categories.join(', ')}`);
        console.log();
    }

    /**
     * Close repository
     */
    close(): void {
        this.repository.close();
    }
}

// CLI execution
async function main() {
    const args = process.argv.slice(2);
    const inputFile = args[0] || path.join(__dirname, '../../data/modules-input.json');

    // Check if file exists
    if (!fs.existsSync(inputFile)) {
        console.error(`❌ File not found: ${inputFile}`);
        console.error('\nUsage: npm run import-modules [path/to/modules.json]');
        console.error('\nExpected JSON format:');
        console.error(JSON.stringify({
            modules: [
                {
                    id: "app:ModuleName",
                    name: "Module Name",
                    app: "App Name",
                    type: "action",
                    description: "What this module does",
                    category: "Category Name",
                    parameters: [
                        {
                            name: "param1",
                            type: "text",
                            required: true,
                            description: "Parameter description"
                        }
                    ]
                }
            ]
        }, null, 2));
        process.exit(1);
    }

    const importer = new ModuleImporter();

    try {
        await importer.importFromFile(inputFile);
        importer.close();
        process.exit(0);
    } catch (error) {
        console.error('\n❌ Import failed:', error);
        importer.close();
        process.exit(1);
    }
}

if (require.main === module) {
    main();
}

export { ModuleImporter };

File: make-mcp/scripts/import_modules.py
#!/usr/bin/env python3
"""
Python-based Module Importer
No compilation required - works on any platform!
"""

import json
import sqlite3
import os
import sys
from pathlib import Path
from typing import List, Dict, Any
import uuid

class ModuleImporter:
    def __init__(self, db_path: str):
        self.db_path = db_path
        self.conn = None
        self.stats = {
            'total': 0,
            'imported': 0,
            'updated': 0,
            'failed': 0,
            'errors': []
        }
        
    def connect(self):
        """Connect to database and create tables if needed"""
        os.makedirs(os.path.dirname(self.db_path), exist_ok=True)
        self.conn = sqlite3.connect(self.db_path)
        self.create_tables()
        
    def create_tables(self):
        """Create database schema"""
        cursor = self.conn.cursor()
        
        # Create modules table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS make_modules (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                module_name TEXT UNIQUE NOT NULL,
                module_type TEXT NOT NULL,
                app_name TEXT NOT NULL,
                app_slug TEXT NOT NULL,
                description TEXT,
                documentation TEXT,
                category TEXT,
                popularity_score INTEGER DEFAULT 50,
                is_premium INTEGER DEFAULT 0,
                icon_url TEXT,
                documentation_url TEXT,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
            )
        ''')
        
        # Create parameters table
        cursor.execute('''
            CREATE TABLE IF NOT EXISTS make_parameters (
                id TEXT PRIMARY KEY,
                module_id INTEGER NOT NULL,
                parameter_name TEXT NOT NULL,
                parameter_type TEXT NOT NULL,
                is_required INTEGER DEFAULT 0,
                description TEXT,
                default_value TEXT,
                options TEXT,
                created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (module_id) REFERENCES make_modules(id) ON DELETE CASCADE
            )
        ''')
        
        # Create FTS5 search index
        cursor.execute('''
            CREATE VIRTUAL TABLE IF NOT EXISTS make_modules_fts USING fts5(
                module_name,
                app_name,
                description,
                documentation,
                category,
                content='make_modules',
                content_rowid='id'
            )
        ''')
        
        # Create triggers for FTS sync
        cursor.execute('''
            CREATE TRIGGER IF NOT EXISTS make_modules_ai AFTER INSERT ON make_modules BEGIN
                INSERT INTO make_modules_fts(rowid, module_name, app_name, description, documentation, category)
                VALUES (new.id, new.module_name, new.app_name, new.description, new.documentation, new.category);
            END
        ''')
        
        cursor.execute('''
            CREATE TRIGGER IF NOT EXISTS make_modules_ad AFTER DELETE ON make_modules BEGIN
                INSERT INTO make_modules_fts(make_modules_fts, rowid, module_name, app_name, description, documentation, category)
                VALUES('delete', old.id, old.module_name, old.app_name, old.description, old.documentation, old.category);
            END
        ''')
        
        cursor.execute('''
            CREATE TRIGGER IF NOT EXISTS make_modules_au AFTER UPDATE ON make_modules BEGIN
                INSERT INTO make_modules_fts(make_modules_fts, rowid, module_name, app_name, description, documentation, category)
                VALUES('delete', old.id, old.module_name, old.app_name, old.description, old.documentation, old.category);
                INSERT INTO make_modules_fts(rowid, module_name, app_name, description, documentation, category)
                VALUES (new.id, new.module_name, new.app_name, new.description, new.documentation, new.category);
            END
        ''')
        
        self.conn.commit()
        
    def slugify(self, text: str) -> str:
        """Convert text to slug"""
        import re
        text = text.lower()
        text = re.sub(r'[^\w\s-]', '', text)
        text = re.sub(r'\s+', '-', text)
        text = re.sub(r'-+', '-', text)
        return text.strip('-')
        
    def calculate_popularity(self, module: Dict) -> int:
        """Calculate module popularity score"""
        score = 50
        
        # Core modules
        if module['app'] in ['HTTP', 'Webhooks', 'Tools']:
            score += 30
            
        # Popular apps
        popular_apps = ['Google Sheets', 'Gmail', 'Slack', 'Airtable', 'Notion']
        if module['app'] in popular_apps:
            score += 20
            
        # Triggers are important
        if module['type'] == 'trigger':
            score += 10
            
        return min(100, score)
        
    def generate_documentation(self, module: Dict) -> str:
        """Generate module documentation"""
        doc = f"# {module['name']}\n\n"
        doc += f"{module['description']}\n\n"
        doc += f"## App\n{module['app']}\n\n"
        doc += f"## Type\n{module['type']}\n\n"
        doc += "## Parameters\n"
        
        for param in module['parameters']:
            required = "(required)" if param['required'] else "(optional)"
            doc += f"- **{param['name']}** {required}: {param.get('description', 'No description')}"
            if 'options' in param:
                doc += f" Options: {', '.join(param['options'])}"
            doc += "\n"
            
        doc += f"\n## Category\n{module['category']}\n"
        return doc
        
    def import_module(self, module: Dict):
        """Import a single module"""
        cursor = self.conn.cursor()
        
        # Check if module exists
        cursor.execute('SELECT id FROM make_modules WHERE module_name = ?', (module['id'],))
        existing = cursor.fetchone()
        is_update = existing is not None
        
        # Insert or update module
        cursor.execute('''
            INSERT INTO make_modules (
                module_name, module_type, app_name, app_slug,
                description, documentation, category, popularity_score,
                is_premium, icon_url, documentation_url, updated_at
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP)
            ON CONFLICT(module_name) DO UPDATE SET
                module_type = excluded.module_type,
                app_name = excluded.app_name,
                app_slug = excluded.app_slug,
                description = excluded.description,
                documentation = excluded.documentation,
                category = excluded.category,
                popularity_score = excluded.popularity_score,
                is_premium = excluded.is_premium,
                icon_url = excluded.icon_url,
                documentation_url = excluded.documentation_url,
                updated_at = CURRENT_TIMESTAMP
        ''', (
            module['id'],
            module['type'],
            module['app'],
            self.slugify(module['app']),
            module['description'],
            self.generate_documentation(module),
            module['category'],
            self.calculate_popularity(module),
            1 if module.get('is_premium', False) else 0,
            module.get('icon_url'),
            module.get('documentation_url')
        ))
        
        module_id = cursor.lastrowid if not is_update else existing[0]
        
        if is_update:
            self.stats['updated'] += 1
            print(f"  ℹ️  Updated existing module")
            # Delete old parameters
            cursor.execute('DELETE FROM make_parameters WHERE module_id = ?', (module_id,))
        
        # Insert parameters
        for param in module['parameters']:
            param_id = str(uuid.uuid4())
            cursor.execute('''
                INSERT INTO make_parameters (
                    id, module_id, parameter_name, parameter_type,
                    is_required, description, default_value, options
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            ''', (
                param_id,
                module_id,
                param['name'],
                param['type'],
                1 if param['required'] else 0,
                param.get('description', ''),
                json.dumps(param.get('default_value')) if 'default_value' in param else None,
                json.dumps(param.get('options')) if 'options' in param else None
            ))
        
        print(f"  📝 Imported {len(module['parameters'])} parameters")
        self.conn.commit()
        
    def import_from_file(self, file_path: str):
        """Import modules from JSON file"""
        print(f"\n🔄 Starting module import from: {file_path}\n")
        
        # Read file
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)
            
        if 'modules' not in data or not isinstance(data['modules'], list):
            raise ValueError('Invalid format: expected { modules: [...] }')
            
        self.stats['total'] = len(data['modules'])
        print(f"📊 Found {self.stats['total']} modules to import\n")
        
        # Import each module
        for i, module in enumerate(data['modules'], 1):
            print(f"[{i}/{self.stats['total']}] Processing: {module['name']} ({module['id']})")
            
            try:
                self.import_module(module)
                self.stats['imported'] += 1
                print(f"  ✅ Imported successfully\n")
            except Exception as e:
                self.stats['failed'] += 1
                error_msg = f"Failed to import {module['id']}: {e}"
                self.stats['errors'].append(error_msg)
                print(f"  ❌ {error_msg}\n")
                
        self.print_summary()
        
    def print_summary(self):
        """Print import summary"""
        print("\n" + "=" * 60)
        print("📊 IMPORT SUMMARY")
        print("=" * 60)
        print(f"Total modules:     {self.stats['total']}")
        print(f"✅ Imported:       {self.stats['imported']}")
        print(f"🔄 Updated:        {self.stats['updated']}")
        print(f"❌ Failed:         {self.stats['failed']}")
        print("=" * 60)
        
        if self.stats['errors']:
            print("\n❌ ERRORS:")
            for error in self.stats['errors']:
                print(f"  - {error}")
                
        print("\n✨ Import complete!\n")
        
        # Database stats
        cursor = self.conn.cursor()
        cursor.execute('SELECT COUNT(*) FROM make_modules')
        total_modules = cursor.fetchone()[0]
        cursor.execute('SELECT COUNT(DISTINCT app_name) FROM make_modules')
        total_apps = cursor.fetchone()[0]
        cursor.execute('SELECT COUNT(*) FROM make_parameters')
        total_params = cursor.fetchone()[0]
        
        print("📈 DATABASE STATISTICS:")
        print(f"  Total modules:   {total_modules}")
        print(f"  Total apps:      {total_apps}")
        print(f"  Total params:    {total_params}")
        print()
        
    def close(self):
        """Close database connection"""
        if self.conn:
            self.conn.close()

def main():
    if len(sys.argv) > 1:
        input_file = sys.argv[1]
    else:
        # Default path
        script_dir = Path(__file__).parent.parent
        input_file = script_dir / 'data' / 'modules-input.json'
        
    if not os.path.exists(input_file):
        print(f"❌ File not found: {input_file}")
        print("\nUsage: python import_modules.py [path/to/modules.json]")
        sys.exit(1)
        
    # Database path
    script_dir = Path(__file__).parent.parent
    db_path = script_dir / 'database' / 'make.db'
    
    importer = ModuleImporter(str(db_path))
    
    try:
        importer.connect()
        importer.import_from_file(str(input_file))
        importer.close()
        sys.exit(0)
    except Exception as e:
        print(f"\n❌ Import failed: {e}")
        import traceback
        traceback.print_exc()
        importer.close()
        sys.exit(1)

if __name__ == '__main__':
    main()

